File: 프로그래머스/lv2/42839. 소수 찾기/소수 찾기.js
function solution(numbers) {
    var answer = 0;
    let flag =0;
    let visited = Array(numbers.length).fill(0);
    let set = new Set();
    function check(n){
        if(n == 1){return false;}
        for(let i=2;i<=Math.sqrt(n);i++){
            if(n%i==0){return false;}
        }
        return true;
    }//소수 판별 함수
    
    function dfs(k){
        set.add(parseInt(k));

        for(let i in numbers){
            if(visited[i]==0){
                visited[i] =1;
                dfs(k+numbers[i]);
                visited[i] =0;
            }
        }
    }
    
    for(let i in numbers){
        if(numbers[i] != '0'){
            visited[i] =1;
            dfs(numbers[i]);
            visited[i] =0;
        }
    }
    
    for(let i of set){
        if(check(i)){answer++;}
    }
    return answer;
}

File: 프로그래머스/2/42839. 소수 찾기/소수 찾기.py
from collections import deque
import math

def solution(numbers):

    N = len(numbers)
    visited = [0] *N
#     bfs로는 백트래킹이 어렵나

    def checkPrime(num):
        num = int(''.join(num))
        if num <= 1:
            return False
        if num == 2:
            return True
        for i in range(2,math.ceil(num**(1/2))+1):
            if num % i == 0:
                return False
        return True
    
    answer = set()
    def dfs(num):
        if checkPrime(num):
            answer.add(int(''.join(num)))
        for i in range(N):
            if not visited[i]:
                visited[i] = 1
                dfs(num + [numbers[i]])
                visited[i] = 0
                
    for i in range(N):
        visited[i] = 1
        dfs([numbers[i]])
        visited[i] = 0

    return len(answer)